x-pack/plugins/actions/server/sub_action_framework/case.ts

```diff
@@ -5,50 +5,45 @@
  * 2.0.
  */
 
-import { schema } from '@kbn/config-schema';
-import {
-  ExternalServiceIncidentResponse,
-  PushToServiceParams,
-  PushToServiceResponse,
-} from './types';
+import { schema, Type } from '@kbn/config-schema';
+import { ExternalServiceIncidentResponse, PushToServiceResponse } from './types';
 import { SubActionConnector } from './sub_action_connector';
 import { ServiceParams } from './types';
 
-export interface CaseConnectorInterface {
-  addComment: ({
-    incidentId,
-    comment,
-  }: {
-    incidentId: string;
-    comment: string;
-  }) => Promise<unknown>;
-  createIncident: (incident: Record<string, unknown>) => Promise<ExternalServiceIncidentResponse>;
+export interface CaseConnectorInterface<Incident, GetIncidentResponse> {
+  addComment: ({ incidentId, comment }: { incidentId: string; comment: string }) => Promise<void>;
+  createIncident: (incident: Incident) => Promise<ExternalServiceIncidentResponse>;
   updateIncident: ({
     incidentId,
     incident,
   }: {
     incidentId: string;
-    incident: Record<string, unknown>;
+    incident: Incident;
   }) => Promise<ExternalServiceIncidentResponse>;
-  getIncident: ({ id }: { id: string }) => Promise<unknown>;
-  pushToService: (params: PushToServiceParams) => Promise<PushToServiceResponse>;
+  getIncident: ({ id }: { id: string }) => Promise<GetIncidentResponse>;
+  pushToService: (params: {
+    incident: { externalId: string | null } & Incident;
+    comments: Array<{ commentId: string; comment: string }>;
+  }) => Promise<PushToServiceResponse>;
 }
 
-export abstract class CaseConnector<Config, Secrets>
+export abstract class CaseConnector<Config, Secrets, Incident, GetIncidentResponse>
   extends SubActionConnector<Config, Secrets>
-  implements CaseConnectorInterface
+  implements CaseConnectorInterface<Incident, GetIncidentResponse>
 {
-  constructor(params: ServiceParams<Config, Secrets>) {
+  constructor(
+    params: ServiceParams<Config, Secrets>,
+    pushToServiceIncidentParamsSchema: Record<string, Type<unknown> | null>
+  ) {
     super(params);
 
     this.registerSubAction({
       name: 'pushToService',
       method: 'pushToService',
       schema: schema.object({
-        incident: schema.object(
-          { externalId: schema.nullable(schema.string()) },
-          { unknowns: 'allow' }
-        ),
+        incident: schema
+          .object({ externalId: schema.nullable(schema.string()) })
+          .extends(pushToServiceIncidentParamsSchema),
```

Comment:
About the test, I was thinking of testing the `pushToServiceIncidentParamsSchema` not the data. For example, if I set `const pushToServiceIncidentParamsSchema = { externalId: schema.number }` it will not override our `{ externalId: schema.nullable(schema.string()) }`.

Reply:
ahh, got it